File: src/main/java/com/nerdearla/workshop/qr/QRService.java
package com.nerdearla.workshop.qr;

import com.nerdearla.workshop.qr.client.QRClient;
import com.nerdearla.workshop.qr.validator.QRValidator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class QRService {

    private static final Logger LOGGER = LoggerFactory.getLogger(QRService.class);

    private final QRClient client;
    private final QRValidator validator;

    public QRService(QRClient qrClient, QRValidator validator) {
        this.client = qrClient;
        this.validator = validator;
    }

    public QR getBy(String qrId) {
        QR qr = client.getById(qrId);
        LOGGER.info("qr found: {}", qr.toString());
        validator.validate(qr);
        LOGGER.info("qr {} validated", qr.getId());
        return qr;
    }
}

File: src/main/kotlin/com/nerdearla/workshop/qr/QRService.kt
package com.nerdearla.workshop.qr

import com.nerdearla.workshop.qr.client.QrClient
import com.nerdearla.workshop.qr.validator.QrValidator
import com.nerdearla.workshop.shared.utils.CompanionLogger
import org.springframework.stereotype.Service

@Service
class QRService(
    private val client: QrClient,
    private val validator: QrValidator
) {
    fun getBy(id: String) =
        doGet(id)
            .also { it.validate() }

    private fun doGet(id: String) =
        client.getBy(id)
            .log { info("qr found: {}", it) }

    private fun QR.validate() =
        validator.validate(this)
            .log { info("qr {} validated", id) }

    companion object : CompanionLogger()
}
